### PR TITLE
OCC-14: Refactor static pages

### PR DESCRIPTION
**Pull request by nlubisch**
**Currently WIP, so don't merge!**

### CODE REVIEW

File: htdocs/app/Migrations/Version20170516084212.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Application\Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration;
+use Doctrine\DBAL\Schema\Schema;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170516084212.php]: please remove phpdoc

File: htdocs/app/Migrations/Version20170516084212.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        $this->addSql('ALTER TABLE page_block ADD FOREIGN KEY (page_id) REFERENCES page(id);');
+    }
+
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function down(Schema $schema)
+    {
+        // this down() migration is auto-generated, please modify it to your needs
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170516084212.php]: please remove this auto generated comment

File: htdocs/app/Migrations/Version20170526212910.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Application\Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration;
+use Doctrine\DBAL\Schema\Schema;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170526212910.php]: lease remove this auto generated comment

File: htdocs/app/Migrations/Version20170526212910.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * @param Schema $schema
+     *
+     * @return void
+     */
+    public function down(Schema $schema)
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170526212910.php]: lease remove this auto generated comment and unneeded break

File: htdocs/app/Migrations/Version20170830005500.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Application\Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration;
+use Doctrine\DBAL\Schema\Schema;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170830005500.php]: lease remove this auto generated comment

File: htdocs/app/Migrations/Version20170830005500.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        $this->addSql('UPDATE languages SET is_translated = 1 WHERE short IN (\'DE\', \'EN\', \'FR\', \'IT\', \'ES\');');
+    }
+
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function down(Schema $schema)
+    {
+        // this down() migration is auto-generated, please modify it to your needs
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170830005500.php]: please remove this auto generated comment

File: htdocs/app/Migrations/Version20170830005500.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+class Version20170830005500 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     * @return void
+     */
+    public function up(Schema $schema)
+    {
+        $this->addSql('ALTER TABLE languages ADD is_translated TINYINT DEFAULT 0 NOT NULL;');
+        $this->addSql('UPDATE languages SET is_translated = 1 WHERE short IN (\'DE\', \'EN\', \'FR\', \'IT\', \'ES\');');
```
Comment (teiling88) [inline htdocs/app/Migrations/Version20170830005500.php:20]: with this migration, following would be not allowed to generate migrations on the old way

File: htdocs/app/Resources/views/page.html.twig [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{% block stylesheets %}{% endblock %}
+
+{% include 'partials/flash-messages.twig' %}
+
+{% block body %}{% endblock %}
+
+{% block javascripts %}{% endblock %}
+</body>
```
Comment (teiling88) [inline htdocs/app/Resources/views/page.html.twig]: why is this closing body tag needed?

File: htdocs/app/config/services.yml [hunk 9]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (teiling88) [inline htdocs/app/config/services.yml]: ???

File: htdocs/app/config/services_oc.yml [hunk 10]
```diff
@@ -20,3 +20,68 @@ services:
...
+        arguments: ['@oc.general.global_context_factory']
+        tags:
+            - { name: kernel.event_subscriber }
+
+    oc.general.global_context_factory:
+        class: Oc\GlobalContext\GlobalContextFactory
+        arguments: ['@oc.general.provider.language_provider']
+
+    # GlobalContext / Provider
+    oc.general.provider.language_provider:
```
Comment (teiling88) [inline htdocs/app/config/services_oc.yml]: why not oc.global_context.provider.language_provider?

File: htdocs/lib2/logic/user.class.php [hunk 11]
```diff
@@ -626,8 +626,6 @@ public function getAnyChanged()
     // return if successfull (with insert)
     public function save()
     {
-        $bNeedStatpicClear = $this->reUser->getChanged('username');
```
Comment (teiling88) [inline htdocs/lib2/logic/user.class.php:629]: why is this change needed?

File: htdocs/src/AppBundle/Controller/PageController.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+use AppBundle\Entity\PageGroup;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Class PageController
+ *
+ * @package AppBundle\Controller
+ * @author Nick Lubisch <oc@nicklubisch.de>
```
Comment (teiling88) [inline htdocs/src/AppBundle/Controller/PageController.php]: I think we didn't have this tag in oc.de or?
Feel free to create the author section in htdocs/composer.json.

File: htdocs/src/Oc/Country/CountryRepository.php [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+     * Creates a country in the database.
+     *
+     * @param CountryEntity $entity
+     *
+     * @return CountryEntity
+     */
+    public function create(CountryEntity $entity)
+    {
+        if (!$entity->isNew()) {
+            throw new InvalidArgumentException('The entity does already exist.');
```
Comment (teiling88) [inline htdocs/src/Oc/Country/CountryRepository.php]: the entity already exists?

File: htdocs/src/Oc/Util/DbalConnection.php [hunk 14]
```diff
@@ -37,7 +53,7 @@ public static function createDbalConnection(
             $params['password'] = $password;
         }
 
-        if ($port) {
+        if ($port !== null) {
```
Comment (teiling88) [inline htdocs/src/Oc/Util/DbalConnection.php:56]: why? I think the if condition is the same?

File: htdocs/src/Oc/User/UserRepository.php [hunk 15]
```diff
@@ -2,7 +2,172 @@
...
+     * Creates a user in the database.
+     *
+     * @param UserEntity $entity
+     *
+     * @return UserEntity
+     */
+    public function create(UserEntity $entity)
+    {
+        if (!$entity->isNew()) {
+            throw new InvalidArgumentException('The entity does already exist.');
```
Comment (teiling88) [inline htdocs/src/Oc/User/UserRepository.php]: the entity already exists?

File: htdocs/src/Oc/Page/PageRepository.php [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+                $queryBuilder->andWhere($column . ' = ' .  $queryBuilder->createNamedParameter($value));
+            }
+        }
+
+        $statement = $queryBuilder->execute();
+
+        $result = $statement->fetch();
+
+        if ($result === false) {
+            return null;
```
Comment (teiling88) [inline htdocs/src/Oc/Page/PageRepository.php]: should we throw an Exception for this case?

File: htdocs/src/Oc/Page/PageEntity.php [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @param array $data
+     *
+     * @return self
+     */
+    public function fromDatabaseArray(array $data)
+    {
+        $this->id = (int) $data['id'];
+        $this->slug = (string) $data['slug'];
+        $this->metaKeywords = (string) $data['meta_keywords'];
```
Comment (teiling88) [inline htdocs/src/Oc/Page/PageEntity.php]: I think we get only string values from database so the type cast to string is not needed or?

**Review by teiling88 — CHANGES_REQUESTED** [review]

File: htdocs/lib2/logic/user.class.php (hunk 11 above)
Comment (nlubisch) [reply htdocs/lib2/logic/user.class.php:629]: unused variable/code, found while debugging some things

File: htdocs/src/Oc/Util/DbalConnection.php (hunk 14 above)
Comment (nlubisch) [reply htdocs/src/Oc/Util/DbalConnection.php:56]: port number can be an integer '0' and would be evaluated as false.
Scrutinizer told this :tongue:

File: htdocs/app/Migrations/Version20170516084212.php (hunk 1 above)
Comment (nlubisch) [reply htdocs/app/Migrations/Version20170516084212.php]: changed this to a meaningful summary

File: htdocs/app/Migrations/Version20170526212910.php (hunk 3 above)
Comment (nlubisch) [reply htdocs/app/Migrations/Version20170526212910.php]: changed this to a meaningful summary

File: htdocs/app/Migrations/Version20170830005500.php (hunk 5 above)
Comment (nlubisch) [reply htdocs/app/Migrations/Version20170830005500.php]: changed this to a meaningful summary

File: htdocs/app/Resources/views/page.html.twig (hunk 8 above)
Comment (nlubisch) [reply htdocs/app/Resources/views/page.html.twig]: removed

File: htdocs/src/AppBundle/Controller/PageController.php (hunk 12 above)
Comment (nlubisch) [reply htdocs/src/AppBundle/Controller/PageController.php]: Tell me what to do :stuck_out_tongue_closed_eyes: its just my common class file template

File: htdocs/src/Oc/User/UserRepository.php (hunk 15 above)
Comment (nlubisch) [reply htdocs/src/Oc/User/UserRepository.php]: i introduce some custom exceptions for this

File: htdocs/src/Oc/Country/CountryEntity.php [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        return $this->short === null;
+    }
+
+    /**
+     * Sets properties from the database.
+     *
+     * @return array
+     */
+    public function toDatabaseArray()
```
Comment (mirsch) [inline htdocs/src/Oc/Country/CountryEntity.php]: Why is ```toDataBaseArray``` and ```fromDatabaseArray``` implemented in the Entity? IMHO this is a part of storing the entities in Database and should be implemented in the repository. Imagine to store the entity in an xml or json storage - the functions make no sense anymore.

File: htdocs/src/Oc/Country/CountryEntity.php [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+
+        return $this;
+    }
+
+    /**
+     * Sets all properties from array.
+     *
+     * @param array $data
+     */
+    public function fromArray(array $data)
```
Comment (mirsch) [inline htdocs/src/Oc/Country/CountryEntity.php]: from- and toArray is very generic. Why not place them in an abstract class?
You should at least define an interface to make sure all entities have the required functions ```isNew```, ```toArray``` and ```fromArray```

**Review by mirsch — CHANGES_REQUESTED** [review]

File: htdocs/src/Oc/Country/CountryEntity.php (hunk 18 above)
Comment (teiling88) [reply htdocs/src/Oc/Country/CountryEntity.php]: This is a idea of expand-ability. But I think in oc.de context it could be moved to the Repository.

File: htdocs/src/Oc/Country/CountryEntity.php (hunk 19 above)
Comment (nlubisch) [reply htdocs/src/Oc/Country/CountryEntity.php]: I already discussed this with @teiling88 weeks ago and he don't want an abstract class nor interface for entities.

File: htdocs/src/Oc/Country/CountryEntity.php (hunk 18 above)
Comment (nlubisch) [reply htdocs/src/Oc/Country/CountryEntity.php]: I agree with @mirsch so i moved them all to the repository.

File: htdocs/src/Oc/Page/PageRepository.php (hunk 16 above)
Comment (nlubisch) [reply htdocs/src/Oc/Page/PageRepository.php]: The new discussed behaviour, that the repository throws exceptions like RecordNotFoundException, RecordsNotFoundException, RecordNotPersistedException and RecordAlreadyExistsException and the service handels them to return default values like an empty array or null.

File: htdocs/src/Oc/Country/CountryEntity.php (hunk 19 above)
Comment (teiling88) [reply htdocs/src/Oc/Country/CountryEntity.php]: Oh my fault @nlubisch I wanted no Abstracted Class but an Interface would be nice :disappointed_relieved:

File: htdocs/src/Oc/Util/DbalConnection.php (hunk 14 above)
Comment (teiling88) [reply htdocs/src/Oc/Util/DbalConnection.php:56]: but port 0 doesn't work :-*

**Review by teiling88 — APPROVED** [review]
